### PR TITLE
Trit tryte sse42 optimize

### PR DESCRIPTION
**Pull request by marktwtn**
Trit tryte conversion acceleration with SIMD SSE4.2.

Including:
- SIMD SSE4.2 acceleration
- Unit test
- Benchmark
- Threshold value of using SSE4.2 acceleration

### CODE REVIEW

File: common/trinary/BUILD [hunk 1]
```diff
@@ -248,3 +251,12 @@ cc_library(
         "//utils:macros",
     ],
 )
+
+cc_library(
+    name = "trit_tryte_sse42",
+    hdrs = ["trit_tryte_sse42.h"],
+    deps = [
+        ":tryte",
+        "//common:defs".
```
Comment (oopsmonk) [inline common/trinary/BUILD]:
```suggestion
        "//common:defs",
```

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: common/trinary/benchmark/BUILD [hunk 2]
```diff
@@ -0,0 +1,8 @@
+cc_test(
+    name = "bench_trit_tryte",
+    timeout = "short",
```
Comment (oopsmonk) [inline common/trinary/benchmark/BUILD]:
```suggestion
    timeout = "long",
```

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: common/trinary/BUILD (hunk 1 above)
Comment (marktwtn) [reply common/trinary/BUILD]: Done.
Sorry for the careless typo.

File: common/trinary/benchmark/BUILD (hunk 2 above)
Comment (marktwtn) [reply common/trinary/benchmark/BUILD]: Done.

File: common/trinary/tests/test_trit_tryte.c [hunk 3]
```diff
@@ -9,26 +9,47 @@
...
+  trit_t trits[60];
   trit_t exp[] = {EXP};
-  trytes_to_trits(trytes, trits, 4);
-  TEST_ASSERT_EQUAL_MEMORY(exp, trits, 12);
+  trytes_to_trits(trytes, trits, 20);
+  TEST_ASSERT_EQUAL_MEMORY(exp, trits, 60);
 }
 
+#if defined(__SSE4_2__)
+void test_trits_to_trytes_sse42(void) {
```
Comment (semenov-vladyslav) [inline common/trinary/tests/test_trit_tryte.c]: Add random tests to cover various lengths and various input data. For each test data convert trits to trytes and back to trits and compare trits, also convert trytes to trits and back to trytes and compare trytes.

File: common/trinary/trit_tryte_sse42.h [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        _mm_sub_epi8(alphabet_offset, _mm_set_epi8(16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16)));
+    __m128i result = _mm_or_si128(_mm_andnot_si128(cmp_result, result_lt), _mm_and_si128(cmp_result, result_ge));
+    /* Store the tryte result */
+    _mm_storeu_si128((__m128i *)(trytes + i * block_8bit), result);
+  }
+  /* The rest of the trits */
+  for (size_t i = (length / NUMBER_OF_TRITS_IN_A_TRYTE / block_8bit) * block_8bit;
+       i * NUMBER_OF_TRITS_IN_A_TRYTE < length; i++) {
+    int k = 0;
+    int used_trit = i * NUMBER_OF_TRITS_IN_A_TRYTE;
```
Comment (semenov-vladyslav) [inline common/trinary/trit_tryte_sse42.h]:
```suggestion
    size_t used_trit = i * NUMBER_OF_TRITS_IN_A_TRYTE;
```

File: common/trinary/benchmark/bench_trit_tryte.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+#include "common/trinary/trit_tryte.h"
+
+// The minimum and maximum input/output tryte size for perfomance testing
+#define MIN_TRYTE_SIZE 16
+#define MAX_TRYTE_SIZE 2048
```
Comment (oopsmonk) [inline common/trinary/benchmark/bench_trit_tryte.c]: let's make it 2673 trytes, the length of a transaction.
```suggestion
#define MAX_TRYTE_SIZE 2673
```

File: common/trinary/benchmark/bench_trit_tryte.c [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+#include "common/trinary/trit_tryte.h"
+
+// The minimum and maximum input/output tryte size for perfomance testing
```
Comment (oopsmonk) [inline common/trinary/benchmark/bench_trit_tryte.c]:
```suggestion
// The minimum and maximum input/output tryte size for performance testing
```

File: common/trinary/benchmark/bench_trit_tryte.c [hunk 7]
```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2020 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
```
Comment (oopsmonk) [inline common/trinary/benchmark/bench_trit_tryte.c]:
```suggestion
 * https://github.com/iotaledger/iota_common
```

File: common/trinary/trit_tryte_sse42.h [hunk 8]
```diff
@@ -0,0 +1,274 @@
+#ifndef __COMMON_TRIT_TRYTE_SSE42_H_
```
Comment (oopsmonk) [inline common/trinary/trit_tryte_sse42.h]: Please add Copyright, thanks

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: common/trinary/tests/test_trit_tryte.c (hunk 3 above)
Comment (marktwtn) [reply common/trinary/tests/test_trit_tryte.c]: Should we consider the case of N(which is not a triple value) trit conversion?
I know that `iota_common` has handles the case but it is not taken care in the SIMD acceleration pull request.

File: common/trinary/trit_tryte_sse42.h (hunk 8 above)
Comment (marktwtn) [reply common/trinary/trit_tryte_sse42.h]: Done.

File: common/trinary/benchmark/bench_trit_tryte.c (hunk 7 above)
Comment (marktwtn) [reply common/trinary/benchmark/bench_trit_tryte.c]: Done.

File: common/trinary/benchmark/bench_trit_tryte.c (hunk 5 above)
Comment (marktwtn) [reply common/trinary/benchmark/bench_trit_tryte.c]: Done.

File: common/trinary/benchmark/bench_trit_tryte.c (hunk 6 above)
Comment (marktwtn) [reply common/trinary/benchmark/bench_trit_tryte.c]: Done.

File: common/trinary/trit_tryte_sse42.h (hunk 4 above)
Comment (marktwtn) [reply common/trinary/trit_tryte_sse42.h]: Done.

File: common/trinary/tests/test_trit_tryte.c (hunk 3 above)
Comment (marktwtn) [reply common/trinary/tests/test_trit_tryte.c]: And maybe the random tests should be raised by another pull request?

**Review by oopsmonk — DISMISSED** [review]
lgfm

File: common/trinary/tests/test_trit_tryte.c (hunk 3 above)
Comment (oopsmonk) [reply common/trinary/tests/test_trit_tryte.c]: I think `bench_trit_tryte.c` has covered the length between 16 to 2673, we need to think about a better way to do monkey testing.

**Review by oopsmonk — APPROVED** [review]

**Pull request merged**